File: 2018/day02a/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	input := read()

	twos, threes := 0, 0
	counts := [26]byte{}
	for _, id := range input {

		counts = [26]byte{}
		for _, r := range id {
			if r < 'a' || r > 'z' {
				fmt.Fprintf(os.Stderr, "ERROR: not a letter\n")
				continue
			}
			counts[r-'a']++
		}

		has2, has3 := false, false
		for _, c := range counts {
			switch c {
			case 2:
				has2 = true
			case 3:
				has3 = true
			}
		}
		if has2 {
			twos++
		}
		if has3 {
			threes++
		}

	}

	fmt.Printf("Answer: %d\n", twos*threes)
}

func read() (input []string) {
	s := bufio.NewScanner(os.Stdin)
	for s.Scan() {
		input = append(input, s.Text())
	}
	return input
}

File: 2018/day02a/main.rs
use std::io;
use std::io::BufRead;

fn main() {
	let stdin = io::stdin();
	let input = stdin.lock().lines().filter_map(|line| line.ok());

	let mut twos = 0;
	let mut threes = 0;
	for l in input {

		let mut counts = [0; 26];
		for c in l.chars() {
			if c < 'a' || c > 'z' {
				println!("ERROR: not a letter");
				continue;
			}
			counts[(c as usize) - ('a' as usize)] += 1;
		}

		let mut has2 = false;
		let mut has3 = false;
		for c in counts.iter() {
			if *c == 2 {
				has2 = true;
			}
			if *c == 3 {
				has3 = true;
			}
		}
		if has2 {
			twos += 1;
		}
		if has3 {
			threes += 1;
		}

	}

	println!("Answer: {}", twos * threes);
}
